### PR TITLE
Test: prepare for Fedora atomic

### PR DESCRIPTION
**Pull request by dperpeet**
- support qcow2 images instead of tarball (can't extract kernel from atomic image)
- add init iso for atomic image (boot)
- update check-login to work with atomic
- fix check-login to change pam smarter, and also /etc/pam.d/sshd
- add timeout to `expect_reload` in testlib

### CODE REVIEW

**Comment by dperpeet** [comment]
Using libguestfs to update the files during vm-install didn't prove successful. Instead, we copy rpms to the vm (as we always do) and extract the files from there to the filesystem.

**Comment by dperpeet** [comment]
successful tests on atomic:
check-login
check-menu
check-terminal

failing:
check-accounts
check-connection
check-dashboard 
check-dbus
check-journal
check-loopback
check-multi-machine

untested:
check-networking check-pages check-realms check-reauthorize check-roles check-services check-session check-shutdown-restart check-storage check-system-info

**Comment by stefwalter** [comment]
> check-dashboard check-multi-machine

Lack of /var/lib/cockpit? If so, that's a real bug.

> check-dbus check-reauthorize

Need cockpit-test-assets

**Comment by dperpeet** [comment]
> Need cockpit-test-assets

thanks for catching that, test-assets are now included as well

File: test/check-login [hunk 1]
```diff
@@ -77,11 +62,15 @@ class TestLogin(MachineCase):
 
         # Try to login as user with correct password
         login ("user", "abcdefg")
-        b.wait_text("#login-error-message", "Permission denied")
+        if 'atomic' in m.os:
+            b.wait_in_text("body", "connect or authenticate: terminated")
```
Comment (mvollmer) [inline test/check-login]: Could we fix Cockpit to also produce "Permission denied" on Atomic?

File: test/check-journal [hunk 2]
```diff
@@ -26,6 +26,10 @@ class TestJournal(MachineCase):
         b = self.browser
         m = self.machine
 
+        # On atomic systems, we need a hack to change files in /usr/lib/systemd
+        if 'atomic' in m.os:
+            m.execute(command="mount -o remount,rw /usr")
```
Comment (mvollmer) [inline test/check-journal]: Would it make sense to factor this out somehow?  Maybe `self.needs_writable_usr()`.

File: test/testvm.py [hunk 3]
```diff
@@ -662,6 +775,8 @@ def modify(gf):
             else:
                 raise Failure("Unsupported OS %s" % self.os)
 
+        if "atomic" in self.os:
+            self.generate_init_image()
```
Comment (mvollmer) [inline test/testvm.py]: What about making a fedor-22-atomic.bootstrap script?  Would that allow us to move some atomic specific code out of testvm?

File: test/check-journal (hunk 2 above)
Comment (dperpeet) [inline test/check-journal]: makes sense, moved to testvm (Machine class)

File: test/testvm.py (hunk 3 above)
Comment (dperpeet) [inline test/testvm.py]: good point. The bootstrap script now also downloads the image if necessary

File: test/testvm.py [hunk 4]
```diff
@@ -158,7 +188,8 @@ def build(self, args):
 
     def run_setup_script(self, script, args):
         """Prepare a test image further by running some commands in it."""
-        self.start(maintain=True)
+        # run this on the original image, not the one in the run directory
+        self.start(maintain=True, original=True)
```
Comment (mvollmer) [inline test/testvm.py]: Hmm, this looks wrong.  The original image in $TEST_DATA/images should never be changed, except when it is overwritten completely by vm-save.

All activity should happen with the image in ./run.  When starting a VM with the image in ./run, the changes made from within that VM are either saved back to the image (during vm-create and vm-install) or thrown away (during check-*).

File: test/testvm.py [hunk 5]
```diff
@@ -766,6 +823,7 @@ def _start_qemu(self, maintain=False, tty=False, monitor=None):
             self._locate_qemu_kvm(),
             "-m", str(MEMORY_MB),
             "-drive", "if=virtio,file=%s,index=0,serial=ROOT,snapshot=%s" % (drive_to_start, snapshot),
+            "-nographic",
```
Comment (mvollmer) [inline test/testvm.py]: By adding this always, we never get a console with vm-run, right?  I think we should have a way to get into a VM even when network is hosed.

File: test/testvm.py [hunk 6]
```diff
@@ -776,6 +834,8 @@ def _start_qemu(self, maintain=False, tty=False, monitor=None):
             "-kernel", self._image_kernel,
             "-initrd", self._image_initrd
         ]
+        if self.init_image:
+            cmd += ["-cdrom", self.init_image]
```
Comment (mvollmer) [inline test/testvm.py]: Does every boot of Atomic need the init ISO?  Or do we only need it during vm-create?

I was hoping that fedora-atomic-22.bootstrap would end with a useable <flavor>-fedora-atomic-22-x86_64.qcow2 which is ready to run the setup script.

File: test/testvm.py [hunk 7]
```diff
@@ -512,7 +559,7 @@ def run_modify_func(self, modify_func):
         if self.verbose:
             gf.set_trace(1)
         try:
-            gf.add_drive_opts(self._image_image, readonly=False)
+            gf.add_drive_opts(self._image_original, readonly=False)
```
Comment (mvollmer) [inline test/testvm.py]: Also looks wrong.

File: test/testvm.py [hunk 8]
```diff
@@ -550,14 +597,20 @@ def unpack_base(self, modify_func=None):
         tarball = os.path.join(self.test_data, "%s.tar.gz" % (image, ))
 
         if os.path.isfile(bootstrap_script):
-            subprocess.check_call([ bootstrap_script, self._image_image, self.arch ])
+            # this image will be created on top of the new base image
+            if os.path.exists(self._image_image):
+                os.unlink(self._image_image)
+            subprocess.check_call([ bootstrap_script, self._image_original, self.arch ])
```
Comment (mvollmer) [inline test/testvm.py]: Using _image_original is wrong.

File: test/testvm.py [hunk 9]
```diff
@@ -550,14 +597,20 @@ def unpack_base(self, modify_func=None):
...
 
         elif os.path.isfile(image_file):
             """ We have a real image file, use that """
-            self.message("Creating disk copy:", self._image_image)
-            shutil.copyfile(image_file, self._image_image)
+            # this image will be created on top of the new base image
+            if os.path.exists(self._image_image):
+                os.unlink(self._image_image)
+            self.message("Creating disk copy:", self._image_original)
+            shutil.copyfile(image_file, self._image_original)
```
Comment (mvollmer) [inline test/testvm.py]: Using _image_original is wrong.

File: test/testvm.py (hunk 6 above)
Comment (mvollmer) [inline test/testvm.py]: > Does every boot of Atomic need the init ISO? Or do we only need it during vm-create?

I guess I am confused.  Looks like every run does indeed need that ISO image, and f-a-22.bootstrap does indeed produce the expected qcow2.

Can we give the ISO a special name, like "cockpit-fedora-atomic-22-x86_64.iso", and if such a file exists, testvm will put it into the cdrom.  That way we can avoid mentioning atomic in testvm.

Also, vm-save and vm-download should handle the ISO.

File: test/testvm.py [hunk 10]
```diff
@@ -70,6 +92,14 @@ def __init__(self, address=None, flavor=None, system=None, arch=None, verbose=Fa
         self.test_data = os.environ.get("TEST_DATA") or self.test_dir
         self.vm_username = "root"
         self.vm_password = "foobar"
+        self.init_image = None
+        self.install_packages_script = None
+        self.target_install_script = None
+        if "atomic" in self.os:
+            self.init_path = os.path.join(self.test_data, "%s-%s_cloud-init" % (self.os, self.arch))
```
Comment (mvollmer) [inline test/testvm.py]: This needs to be 

```
self.init_path = os.path.join(self.test_data, "images/%s-%s-%s_cloud-init" % (self.flavor, self.os, self.arch))
```

File: test/check-pages [hunk 11]
```diff
@@ -49,8 +49,10 @@ WantedBy=default.target
         b.enter_page("/system/init")
         b.wait_text("#service-unit .panel-heading", "Test Service")
 
-        # check that the login dance preserves parameters
-        m.execute("systemctl restart cockpit-testing.service")
+        # testing service has unmet dependencies on atomic
+        if not 'atomic' in m.os:
+            # check that the login dance preserves parameters
+            m.execute("systemctl restart cockpit-testing.service")
```
Comment (mvollmer) [inline test/check-pages]: `cockpit-testing.servce` is just a variant of `cockpit.service`, it does not start the test-server, so it shouldn't have dependency problems.

File: test/testvm.py (hunk 4 above)
Comment (dperpeet) [inline test/testvm.py]: you're right - I didn't consider our hubbot setup. In my testing environment, I don't clean up ./run before I create a new image.

File: test/testvm.py (hunk 6 above)
Comment (dperpeet) [inline test/testvm.py]: I like the idea of autoloading an iso image via filename magic.

**Comment by mvollmer** [comment]
`check-roles` fails because /var/log/lastlog does not exist.  Adding

```
m.execute("touch var/log/lastlog")
```

at the beginning of testDynamic makes it work.

However, we should change cockpit-accounts.js to survive without /var/log/lastlog.  I think simply not opening the "unexpected error" dialog should be enough.

**Comment by mvollmer** [comment]
#2397 should fix check-roles.

File: test/check-login (hunk 1 above)
Comment (dperpeet) [inline test/check-login]: This probably relates to using ssh under the hood. I agree: a proper error message would be good here.

Comment (dperpeet) [inline test/check-login]: follow-up issue https://github.com/cockpit-project/cockpit/issues/2402

**Comment by dperpeet** [comment]
all previous comments addressed or posted as follow-up

File: test/testlib.py [hunk 12]
```diff
@@ -414,16 +427,56 @@ def tearDown(self):
...
+
+    def start_cockpit(self, atomic_wait_for_host="localhost"):
         """Start Cockpit.
 
         Cockpit is not running when the test virtual machine starts up, to
         allow you to make modifications before it starts.
         """
-        self.machine.execute("systemctl start cockpit-testing.socket")
+        if "atomic" in self.machine.os:
+            # HACK
```
Comment (stefwalter) [inline test/testlib.py]: @dperpeet Have you filed a bug about this upstream? If so, can you include the link here?

Comment (dperpeet) [inline test/testlib.py]: done

**Pull request closed without merging**